### ISSUE TITLE
on2fa prompts multiple times

### ISSUE THREAD

**Issue by ruxandrafed** [octokit/octokit.js#1279]
Hi guys,

I updated a CLI project I was working on to use the latest version, and I'm running into some odd issues re. `on2fa`.

Inside a separate function, I am creating the client like this:
```
const createAuthedClient = async (credentials) => {
  try {
    const client = new Octokit({
      auth: {
        username: credentials.username,
        password: credentials.password,
        async on2fa () {
          const { otp } = await inquirer.askGitHubOTP()
          return otp
        }
      }
    })
    return client
  } catch (error) {
    throw error
  }
}
```

and `inquirer.askGitHubOTP` only does the following:
```
const askGitHubOTP = () => {
  const questions = [
    {
      name: 'otp',
      type: 'input',
      message: 'Your GitHub OTP/2FA Code:',
      filter: input => input.trim()
    }
  ]
  return inquirer.prompt(questions)
}
```

I only run this once in my flow, but for some reason, the `on2fa` method ends up executing 4 times, so I get prompted 4 times for the code. As long as I type the right one, it keeps going .. I must say, also, that it occasionally happened to prompt me just once or twice.

What am I doing wrong? Any tips? Thank you!!

**Comment by gr2m** [comment]
Hi Ruxandra 👋 

I don’t see any problem right away 🤔 Is the full code somewhere I could have a look at it? If not, could you set up a minimal repository with only the on2fa handling + inquirer?

Can you set logging as described at https://github.com/octokit/rest.js#debug, maybe it gives us some more information.

**Comment by ruxandrafed** [comment]
Hey Gregor, thanks for the prompt response!

I set up logging and noticed that the `on2fa` actually seems to run when I use other authorization methods. Let me take a step back and explain what this CLI is trying to do:

1. Prompt user how they want to authenticate - token or user/pass
2. If user picks user/pass and has 2FA, get credentials and 2FA code
3. Create the client using the credentials.
4. Generate a token with the `repo` and `read:org` scopes (using [createAuthorization](https://octokit.github.io/rest.js/#api-OauthAuthorizations-createAuthorization))
5. If the token already exists, list all authorizations (using [listAuthorizations](https://octokit.github.io/rest.js/#api-OauthAuthorizations-listAuthorizations)) and find the auth id for it
6. Delete the old authorization (using [deleteAuthorization](https://octokit.github.io/rest.js/#api-OauthAuthorizations-deleteAuthorization)) and get back to step 3.
7. Once the token was successfully created, create the client again using token authentication.
8. Later, have the ability to push to a repository.

Unfortunately, the repository I'm working on is private, so I can't easily share it with you, but hoping this ^ can help. I must be doing something wrong 😄Please advise, thank you!

Here is what running these in debug mode logged:

```
? GitHub OTP/2FA Code: 242434
request { method: 'POST',
  baseUrl: 'https://api.github.com',
  headers:
   { accept: 'application/vnd.github.v3+json',
     'user-agent': 'octokit.js/16.17.1 Node.js/8.11.3 (macOS High Sierra; x64)',
     'x-github-otp': '242434',
     authorization: 'Basic <redacted>',
     'content-type': 'application/json; charset=utf-8' },
  request:
   { validate:
      { client_id: [Object],
        client_secret: [Object],
        fingerprint: [Object],
        note: [Object],
        note_url: [Object],
        scopes: [Object] } },
  url: '/authorizations',
  scopes: [ 'read:org', 'repo' ],
  note: '<redacted>',
  note_url: '<redacted>' }
POST /authorizations - 422 in 400ms
request { method: 'GET',
  baseUrl: 'https://api.github.com',
  headers:
   { accept: 'application/vnd.github.v3+json',
     'user-agent': 'octokit.js/16.17.1 Node.js/8.11.3 (macOS High Sierra; x64)',
     authorization: 'Basic <redacted>' },
  request: { validate: { page: [Object], per_page: [Object] } },
  url: '/authorizations',
  per_page: 100 }
GET /authorizations?per_page=100 - 401 in 1083ms

? GitHub OTP/2FA Code: 559147
request { method: 'GET',
  baseUrl: 'https://api.github.com',
  headers:
   { accept: 'application/vnd.github.v3+json',
     'user-agent': 'octokit.js/16.17.1 Node.js/8.11.3 (macOS High Sierra; x64)',
     'x-github-otp': '559147',
     authorization: 'Basic <redacted>' },
  request: { validate: { page: [Object], per_page: [Object] } },
  url: '/authorizations',
  per_page: 100 }
GET /authorizations?per_page=100 - 200 in 471ms
request { method: 'DELETE',
  baseUrl: 'https://api.github.com',
  headers:
   { accept: 'application/vnd.github.v3+json',
     'user-agent': 'octokit.js/16.17.1 Node.js/8.11.3 (macOS High Sierra; x64)',
     authorization: 'Basic <redacted>' },
  request: { validate: { authorization_id: [Object] } },
  url: '/authorizations/:authorization_id',
  authorization_id: 269748818 }
DELETE /authorizations/269748818 - 401 in 363ms

? GitHub OTP/2FA Code: 922500
request { method: 'DELETE',
  baseUrl: 'https://api.github.com',
  headers:
   { accept: 'application/vnd.github.v3+json',
     'user-agent': 'octokit.js/16.17.1 Node.js/8.11.3 (macOS High Sierra; x64)',
     'x-github-otp': '922500',
     authorization: 'Basic <redacted>=',
     'content-length': 0 },
  request: { validate: { authorization_id: [Object] } },
  url: '/authorizations/:authorization_id',
  authorization_id: 269748818 }
DELETE /authorizations/269748818 - 204 in 412ms
request { method: 'POST',
  baseUrl: 'https://api.github.com',
  headers:
   { accept: 'application/vnd.github.v3+json',
     'user-agent': 'octokit.js/16.17.1 Node.js/8.11.3 (macOS High Sierra; x64)',
     authorization: 'Basic <redacted>=' },
  request:
   { validate:
      { client_id: [Object],
        client_secret: [Object],
        fingerprint: [Object],
        note: [Object],
        note_url: [Object],
        scopes: [Object] } },
  url: '/authorizations',
  scopes: [ 'read:org', 'repo' ],
  note: '<redacted>',
  note_url: '<redacted>' }
POST /authorizations - 401 in 386ms

? GitHub OTP/2FA Code: 922500
request { method: 'POST',
  baseUrl: 'https://api.github.com',
  headers:
   { accept: 'application/vnd.github.v3+json',
     'user-agent': 'octokit.js/16.17.1 Node.js/8.11.3 (macOS High Sierra; x64)',
     'x-github-otp': '922500',
     authorization: 'Basic <redacted>=',
     'content-type': 'application/json; charset=utf-8' },
  request:
   { validate:
      { client_id: [Object],
        client_secret: [Object],
        fingerprint: [Object],
        note: [Object],
        note_url: [Object],
        scopes: [Object] } },
  url: '/authorizations',
  scopes: [ 'read:org', 'repo' ],
  note: '<redacted>',
  note_url: '<redacted>' }
POST /authorizations - 201 in 469ms
```

**Comment by gr2m** [comment]
The process reads legit, in fact this should be built-in into Octokit 👍 

So what I think the problem is that octokit does send the OTP code only once, it does not keep it in its state and keeps sending it. You see the 2nd request to `GET /authorizations?per_page=100` does not send the `x-github-otp` header. That is what we need to fix.

Would you like to try to fix it? You could start out by sending a pull request with a failing test and we take it from there? The test would go into https://github.com/octokit/rest.js/tree/master/test/issues

**Comment by ruxandrafed** [comment]
Yey, so it does make sense! I'll have a look 😄 Thanks, @gr2m!

**Comment by ruxandrafed** [comment]
@gr2m Opened #1281 with the failing test. Had a quick look at how I'd fix it, but .. I'm not there yet 😄 Some guidance would be welcome, thanks!

**Comment by ruxandrafed** [comment]
@gr2m would this be a good approach? Thank you!

```diff
Index: plugins/authentication/before-request.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- plugins/authentication/before-request.js	(date 1552539831000)
+++ plugins/authentication/before-request.js	(date 1552540404000)
@@ -3,6 +3,10 @@
 const btoa = require('btoa-lite')
 
 function authenticationBeforeRequest (state, options) {
+  if (state.otp) {
+    options.headers['x-github-otp'] = state.otp
+  }
+
   if (typeof state.auth === 'string') {
     options.headers['authorization'] = state.auth
 
Index: plugins/authentication/request-error.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- plugins/authentication/request-error.js	(date 1552539831000)
+++ plugins/authentication/request-error.js	(date 1552540329000)
@@ -27,6 +27,7 @@
       const newOptions = Object.assign(options, {
         headers: Object.assign({ 'x-github-otp': oneTimePassword }, options.headers)
       })
+      state.otp = oneTimePassword
       return state.octokit.request(newOptions)
     })
 }
```

**Comment by ruxandrafed** [comment]
Fixed with #1281. Thank you for merging! Will follow up with another PR for https://github.com/octokit/rest.js/pull/1281#issuecomment-472710162.

**Comment by gr2m** [comment]
@ruxandrafed I found out by chance that you can you can filter results from [`octokit.oauthAuthorizations.listAuthorizations`]( https://octokit.github.io/rest.js/#api-OauthAuthorizations-listAuthorizations) by passing a `client_id` option. `00000000000000000000` is the client id for personal access tokens. 

```js
  const { data: authorizations } = await octokit.oauthAuthorizations.listAuthorizations({
    client_id: '00000000000000000000'
  })
```

This will make your `5. If the token already exists, list all authorizations (using listAuthorizations) and find the auth id for it` step a little more efficient. 

I’m looking if there is an even easier way to reset a token if there is a conflict with the provided `note`. I want to add the reset feature directly into Octokit :)

**Comment by ruxandrafed** [comment]
Thanks @gr2m! I went through the docs and stumbled upon that, however in my case it doesn't help much because I don't know the client id. If I could retrieve one by note, though.. that would certainly make my life easier 😊

Let me know if you need a helping hand for Octokit!
